File: wear/src/main/java/com/example/next_contest/wear/util/GeoUtils.java
package com.example.next_contest.wear.util;

import android.location.Location;

public final class GeoUtils {
    private GeoUtils() {
    }

    public static float distanceMeters(
            double startLatitude,
            double startLongitude,
            double endLatitude,
            double endLongitude
    ) {
        float[] result = new float[1];
        Location.distanceBetween(
                startLatitude,
                startLongitude,
                endLatitude,
                endLongitude,
                result
        );
        return result[0];
    }

    public static float bearing(
            double startLatitude,
            double startLongitude,
            double endLatitude,
            double endLongitude
    ) {
        double startLatRad = Math.toRadians(startLatitude);
        double endLatRad = Math.toRadians(endLatitude);
        double deltaLngRad = Math.toRadians(endLongitude - startLongitude);

        double y = Math.sin(deltaLngRad) * Math.cos(endLatRad);
        double x = Math.cos(startLatRad) * Math.sin(endLatRad)
                - Math.sin(startLatRad) * Math.cos(endLatRad) * Math.cos(deltaLngRad);

        double bearing = Math.toDegrees(Math.atan2(y, x));
        return (float) ((bearing + 360.0) % 360.0);
    }

    public static String bearingToDirectionText(float bearing) {
        String[] directions = {
                "북쪽",
                "북동쪽",
                "동쪽",
                "남동쪽",
                "남쪽",
                "남서쪽",
                "서쪽",
                "북서쪽"
        };

        int index = (int) ((bearing + 22.5f) / 45.0f) % directions.length;
        return directions[index];
    }
}

File: app/src/main/java/com/example/next_contest/util/GeoUtils.kt
package com.example.next_contest.util

import android.location.Location

object GeoUtils {

    fun getBearing(
        lat1: Double,
        lng1: Double,
        lat2: Double,
        lng2: Double
    ): Float {
        val dLon = Math.toRadians(lng2 - lng1)

        val y = Math.sin(dLon) * Math.cos(Math.toRadians(lat2))

        val x = Math.cos(Math.toRadians(lat1)) * Math.sin(Math.toRadians(lat2)) -
                Math.sin(Math.toRadians(lat1)) *
                Math.cos(Math.toRadians(lat2)) *
                Math.cos(dLon)

        val bearing = Math.toDegrees(Math.atan2(y, x))

        return ((bearing + 360) % 360).toFloat()
    }

    fun distanceBetween(
        lat1: Double,
        lng1: Double,
        lat2: Double,
        lng2: Double
    ): Float {
        val results = FloatArray(1)
        Location.distanceBetween(lat1, lng1, lat2, lng2, results)
        return results[0]
    }
}